10up/distributor
# When transferring all in-content images, ensure any image URLs in the content are updated to the new image

### Description of the Change

We currently have two options for processing media when content is distributed:

1. Only send the featured image
2. Send the featured image plus any attached media

With the second option turned on, all attached images (if using the Classic Editor) or any block image (if using the Block Editor) will be transferred to the destination site. But we don't actually update any of the in-content URLs, so those will still point to the origin site.

This PR fixes that when the second option is turned on. We keep track of any images we distribute and we then parse through the content of the newly created item to find any old image URLs and update with the new image URLs. This is set up to work both for the Block Editor and Classic Editor.

> [!NOTE]  
> We try and determine the original image size and use that when getting the new image URL. If an image size from the origin site doesn't exist on the destination site, the full size image will be used instead.

#### For the Classic Editor

We parse through the entire post content, using the HTML API, to find all images. We pull out the image classes and use those to match the existing image using the image ID class and the image size using the image size class. If no classes exist or the image ID class doesn't exist, we skip that image since we can't be sure it's a match. Note we could change this logic to try matching based on the image URL but the problem here is we have the original full image URL which is likely not the same URL used in the content (as that's typically a resized image URL).

For any images found that match the image ID, we use the HTML API to update the image src with the image that matches the image size class and update the image ID class.

#### For the Block Editor

We parse the content into blocks and find any image blocks that have the old image ID as a block attribute (similar to above, we could match based on URL but for image blocks, I think we can depend on the image ID always existing and being easier to match on). We update the ID block attribute and then use the HTML API to update the image src with the proper sized image (based on the image size block attribute) and update the image class.

Note that right now we only update image blocks. We could look to add support for other blocks that use images, like the cover block, media and text block, etc.

Closes #1278
Closes #1262
Closes #372

### How to test the Change

1. Ensure the `Media Handling` setting at `wp-admin/admin.php?page=distributor-settings` is set to `Process the featured image and any attached images`
2. Create a new Block Editor post and add at least one image block. Add new images to those blocks (this ensures the image is attached to the post)
3. Push this content to a connection
4. View the content on the destination site and ensure the image URLs reflect that site
5. Run through the same steps but install the Classic Editor plugin on both sites first and create a new classic post
6. Run through all of the above again but Pull content instead of Push
7. If desired, run through the same steps but use an external connection instead of an internal connection

### Changelog Entry

> Added - When the setting to distribute all images is turned on, do a search/replace on the distributed content to update the image URLs from the origin site to the destination site

### Credits

Props @dkotter, @willemb2, @dcarrionc, @lucymtc, @iamdharmesh 

### Checklist:
